google/cloud/storage/internal/sha256_hash_test.cc

```diff
@@ -0,0 +1,55 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/sha256_hash.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+namespace {
+
+TEST(Sha256Hash, HexEncodeEmpty) { EXPECT_EQ("", HexEncode({})); }
+
+TEST(Sha256Hash, HexEncodeBasic) {
+  EXPECT_EQ("0001ff7f10", HexEncode({0x00, 0x01, 0xFF, 0x7F, 0x10}));
```

Comment:
Does the v4 signing require upper or lowercase hex digits? or do either work?

Reply:
It requires lowercase, the algorithm takes `HexEncode(Sha256Hash(complicated stuff))` and then signs that (and some other stuff) with some other hash / encryption algo.  Uppercase in `HexEncode()` would not work, the hash of the hash would mismatch.